### PR TITLE
Update libc and some fixes for x86_64-unknown-linux-gnux32

### PR DESCRIPTION
**Pull request by malbarbo**

### CODE REVIEW

**Comment by malbarbo** [comment]
The libstd tests fails in my machine, let's see on travis.

File: src/libcore/tests/hash/sip.rs [hunk 1]
```diff
@@ -249,12 +249,21 @@ fn test_hash_usize() {
     assert!(hash(&(val as u64)) != hash(&(val as usize)));
     assert_eq!(hash(&(val as u32)), hash(&(val as usize)));
 }
-#[test] #[cfg(target_arch = "x86_64")]
+
+#[test] #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
```
Comment (alexcrichton) [inline src/libcore/tests/hash/sip.rs]: We've got existing tests for x86/x86_64, perhaps those could just switch to being two tests based on target_pointer_width?

File: src/libstd/sys/unix/condvar.rs [hunk 2]
```diff
@@ -92,7 +92,16 @@ impl Condvar {
         assert_eq!(r, 0);
 
         // Nanosecond calculations can't overflow because both values are below 1e9.
-        let nsec = dur.subsec_nanos() as libc::c_long + now.tv_nsec as libc::c_long;
+        // HACK: add support to linux x32, the original code was
```
Comment (alexcrichton) [inline src/libstd/sys/unix/condvar.rs]: Is this necessary b/c the values are always < 1e9? Doesn't that mean we can basically cast to "anything reasonable" here and expect it to work?

File: src/ci/docker/x86_64-gnu/Dockerfile [hunk 3]
```diff
@@ -16,5 +16,17 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 COPY scripts/sccache.sh /scripts/
 RUN sh /scripts/sccache.sh
 
-ENV RUST_CONFIGURE_ARGS --build=x86_64-unknown-linux-gnu --enable-sanitizers --enable-profiler
-ENV SCRIPT python2.7 ../x.py test
+ENV HOSTS=x86_64-unknown-linux-gnu
+
+ENV TARGETS=x86_64-unknown-linux-gnu
+ENV TARGETS=$TARGETS,x86_64-unknown-linux-gnux32
```
Comment (alexcrichton) [inline src/ci/docker/x86_64-gnu/Dockerfile]: Do you plan on leaving this permanently? It looks like we don't have a lot of extra time on this builder unfortunately (the last one clocked in at 1h40m)...

I'd imagine though that once we get the test suite passing once it's probably pretty unlikely to regress?

**Review by alexcrichton — COMMENTED** [review]
Looks great!

**Comment by alexcrichton** [comment]
Looks like maybe something related to stack walking is odd?

```
[01:12:28]     [run-pass] run-pass/backtrace-debuginfo.rs

[01:12:28]     [run-pass] run-pass/backtrace.rs

[01:12:28]     [run-pass] run-pass/command-before-exec.rs

[01:12:28]     [run-pass] run-pass/command-exec.rs

[01:12:28]     [run-pass] run-pass/out-of-stack.rs

[01:12:28]     [run-pass] run-pass/panic-runtime/lto-unwind.rs

[01:12:28]     [run-pass] run-pass/stack-probes-lto.rs

[01:12:28]     [run-pass] run-pass/stack-probes.rs

[01:12:28]     [run-pass] run-pass/wait-forked-but-failed-child.rs
```

File: src/libcore/tests/hash/sip.rs (hunk 1 above)
Comment (malbarbo) [reply src/libcore/tests/hash/sip.rs]: Done.

File: src/libstd/sys/unix/condvar.rs (hunk 2 above)
Comment (malbarbo) [reply src/libstd/sys/unix/condvar.rs]: Should we use i32?

Comment (alexcrichton) [reply src/libstd/sys/unix/condvar.rs]: I think that could work? It just seemed like the extra casting wasn't necessary here given the various constraints.

**Comment by malbarbo** [comment]
@alexcrichton I squashed the commits and disable the gnux32 tests. I also opened the issue https://github.com/rust-lang/rust/issues/45416 to track the failures.

File: src/libstd/sys/unix/condvar.rs (hunk 2 above)
Comment (malbarbo) [reply src/libstd/sys/unix/condvar.rs]: As subsec_nanos is u32, I opted to use u32. Both i32 and u32 are fine.

**Comment by alexcrichton** [comment]
Ok! Is this ready to go? (did you mean to add ignore annotations to all affected tests?)

**Comment by malbarbo** [comment]
I think this is ready to go. This will allows others to test the target with the fixed libc.

I think you are correct about the stack walk odds. I will be out for some time now, but it would be great if someone could take a look at it.

**Comment by Mark-Simulacrum** [comment]
Is this intended to fix https://github.com/rust-lang/rust/issues/45417, https://github.com/rust-lang/rust/issues/45416 and perhaps others?

**Comment by malbarbo** [comment]
> Is this intended to fix #45417, #45416 and perhaps others?

No. :disappointed:

**Pull request merged**